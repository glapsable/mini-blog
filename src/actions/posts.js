export const addPost = (postData = {}) => {
  const {
    title = '',
    createdAt = 0,
    amount = 0,
    description = '',
  } = postData;
  const post = {
    title, createdAt, amount, description,
  };
  return {
    type: 'ADD_POST',
    post,
  };
};

export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id,
});

export const editPost = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
