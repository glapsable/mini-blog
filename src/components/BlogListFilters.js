import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate,
} from '../actions/filters';

class BlogListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  onTextChange = (e) => {
    const { setTextFilterDispatch } = this.props;
    setTextFilterDispatch(e.target.value);
  };

  onSortChange = (e) => {
    const { sortByDateDispatch, sortByAmountDispatch } = this.props;
    const { value } = e.target;
    if (value === 'date') sortByDateDispatch();
    if (value === 'amount') sortByAmountDispatch();
  };

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDateDispatch, setEndDateDispatch } = this.props;
    setStartDateDispatch(startDate);
    setEndDateDispatch(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  };

  render() {
    const { filters } = this.props;
    const { calendarFocused } = this.state;
    return (
      <div>
        <input
          type="text"
          placeholder="Search posts"
          value={filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">By date</option>
          <option value="amount">By amount</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate}
          startDateId="startDateFilter"
          endDate={filters.endDate}
          endDateId="endDateFilter"
          onDatesChange={this.onDatesChange}
          focusedInput={calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates
          numberOfMonths={1}
          isOutsideRange={() => false}
          showDefaultInputIcon
        />
      </div>
    );
  }
}

BlogListFilters.propTypes = {
  filters: PropTypes.shape({
    text: PropTypes.string,
    sortBy: PropTypes.string,
    startDate: PropTypes.object,
    endDate: PropTypes.object,
  }).isRequired,
  setStartDateDispatch: PropTypes.func.isRequired,
  setEndDateDispatch: PropTypes.func.isRequired,
  setTextFilterDispatch: PropTypes.func.isRequired,
  sortByDateDispatch: PropTypes.func.isRequired,
  sortByAmountDispatch: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  filters: state.filters,
});

const mapDispatchToProps = dispatch => ({
  setTextFilterDispatch: text => dispatch(setTextFilter(text)),
  sortByDateDispatch: () => dispatch(sortByDate()),
  sortByAmountDispatch: () => dispatch(sortByAmount()),
  setStartDateDispatch: startDate => dispatch(setStartDate(startDate)),
  setEndDateDispatch: endDate => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogListFilters);
