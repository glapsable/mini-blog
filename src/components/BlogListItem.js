import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

// load a locale
numeral.register('locale', 'pl', {
  delimiters: {
    thousands: ' ',
    decimal: ',',
  },
  abbreviations: {
    thousand: 'k',
    million: 'm',
    billion: 'b',
    trillion: 't',
  },
  currency: {
    symbol: 'PLN',
  },
});

// switch between locales
numeral.locale('pl');

const BlogListItem = ({
  id, title, amount, createdAt, description,
}) => (
  <Link to={`/edit/${id}`}>
    <h3>{title}</h3>
    <p>{description}</p>
    <p>Trip amount: {numeral(amount / 100).format('0,0[.]00 $')}</p>
    <p>{moment(createdAt).format('Do MMMM YYYY')}</p>
  </Link>
);

BlogListItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  amount: PropTypes.number.isRequired,
  createdAt: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
};

export default BlogListItem;
