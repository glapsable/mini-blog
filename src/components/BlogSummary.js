import React from 'react';

const BlogSummary = () => (
  <p>Blog Summary</p>
);

export default BlogSummary;
