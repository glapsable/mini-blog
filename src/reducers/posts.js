const postsReducerDefaultState = [];
export default (state = postsReducerDefaultState, action) => {
  switch (action.type) {
    default:
      return state;
  }
};
