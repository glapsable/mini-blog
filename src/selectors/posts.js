import moment from 'moment';

export default (posts, {
  text, sortBy, startDate, endDate,
}) => posts.filter((post) => {
  const createdAtMoment = moment(post.createdAt);
  const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
  const endDateMatch = endDate ? endDate.isSameOrBefore(createdAtMoment, 'day') : true;
  const textMatch = post.title.toLowerCase().includes(text.toLowerCase());

  return startDateMatch && endDateMatch && textMatch;
}).sort((a, b) => {
  if (sortBy === 'date') return a.createdAt < b.createdAt ? 1 : -1;
  if (sortBy === 'amount') return a.amount < b.amount ? 1 : -1;
  return 0;
});
