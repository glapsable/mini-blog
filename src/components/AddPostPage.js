import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { addPost } from '../actions/posts';
import PostForm from './PostForm';

export class AddPostPage extends React.Component {
  onSubmit = (post) => {
    const { addPostDispatch, history } = this.props;
    addPostDispatch(post);
    history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Add Post Page</h1>
        <PostForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

AddPostPage.propTypes = {
  addPostDispatch: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
};

const mapDispatchToProps = dispatch => ({
  addPostDispatch: post => dispatch(addPost(post)),
});

export default connect(undefined, mapDispatchToProps)(AddPostPage);
