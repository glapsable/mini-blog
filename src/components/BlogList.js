import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import selectPosts from '../selectors/posts';
import BlogListItem from './BlogListItem';

const BlogList = (props) => {
  const { posts } = props;
  return (
    <div>
      {
        posts.length === 0 ? (
          <p>There no posts</p>
        ) : (
          posts.map(post => (
            <BlogListItem {...post} key={post.id} />
          ))
        )
      }
    </div>
  );
};

BlogList.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const mapStateToProps = state => ({
  posts: selectPosts(state.posts, state.filters),
});

export default connect(mapStateToProps)(BlogList);
