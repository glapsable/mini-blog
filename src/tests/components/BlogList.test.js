import React from 'react';
import { shallow } from 'enzyme';
import { BlogList } from '../../components/BlogList';
import posts from '../fixtures/posts';

test('should render BlogList correctly with posts', () => {
  const wrapper = shallow(<BlogList posts={posts} />);
  expect(wrapper).toMatchSnapshot();
});

test('should renders BlogList correctly with no posts', () => {
  const wrapper = shallow(<BlogList posts={[]} />);
  expect(wrapper).toMatchSnapshot();
});
