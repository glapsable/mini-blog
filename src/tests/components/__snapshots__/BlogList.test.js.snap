// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogList correctly with posts 1`] = `
<div>
  <BlogListItem
    amount={0}
    createdAt={0}
    description="aaa"
    id="aaa"
    key="aaa"
    title="aaa"
  />
  <BlogListItem
    amount={1500}
    createdAt={-345600000}
    description="bbb"
    id="bbb"
    key="bbb"
    title="bbb"
  />
  <BlogListItem
    amount={50}
    createdAt={604800000}
    description="ccc"
    id="ccc"
    key="ccc"
    title="ccc"
  />
</div>
`;

exports[`should renders BlogList correctly with no posts 1`] = `
<div>
  <p>
    There no posts
  </p>
</div>
`;
